chargify/active_merchant
# add verify method to bogus

### Why:
**Moneris wrote:**

1. Please note that you CANNOT store CVD value. The example you provided is for adding a cc into the vault.

2. I am not sure if you are aware, there is a new mandate that is coming into effect on April 2020. It is called Credential on File . What this mandate requires is that merchants who transact using stored cards need to send additional fields ( cof object) to indicate if the card is being stored and whether it is an initial transaction or subsequent transaction. In this scenario, the bank will decline due to the lack of cof fields and initial transaction should have cvv values as well.

3. The initial transaction should contain CVV value but you will never store the value. So in your subsequent transactions, your will not be providing the CVV value since the card info provided are stored credentials and cvv cannot be stored. The flow is always like this – Initial >Storage> Subsequent
If you have stored cards in your vault prior to the mandate, they are considered as legacy cards and the process to  bring them up to mandate does not require you to provide CVV and your COF object will have the values ( 2, U, null).


### What:
**Added Card Verification method for tests.** 

* For **CVV Disable** process looks the same like for _CVV Enabled_  but without _CVD params_ for _Card Verification_

* For **CVV Enabled** process: 

1. **Card Verification** with CVD and Credential on File (like a first transaction)
"Merchants can use Card Verification with COF AVS and CVD as the first transaction prior to storing the credentials instead of Purchase or Preauthorization." _source_: [developer.moneris](https://developer.moneris.com/Documentation/NA/E-Commerce%20Solutions/API/Card%20Verification)
- I send CC details.
- I send CVV and I get CVV acceptance
- I send Credential on File and I get IssuerId

2. **Vault Add Credit Card** _source_: [developer.moneris](https://developer.moneris.com/Documentation/NA/E-Commerce%20Solutions/API/Vault)

3. **Purchase** with Vault and Credential on File _source_ [developer.moneris](https://developer.moneris.com/Documentation/NA/E-Commerce%20Solutions/API/Purchase)

Used Credential on File scenario:
![image](https://user-images.githubusercontent.com/8780680/80367834-0e018100-888c-11ea-8335-fa175c0cae58.png)

### Test:
1. Create a site with Moneris payment gateway. 
You can use test settings for Moneris: ([source](https://developer.moneris.com/en/More/Testing/Testing%20a%20Solution))
```
STORE_ID: store5
API_TOKEN: yesguy
CVV Enabled: check <------ IMPORTANT
```
2. Buy subscription and renewal it.
3. Visit [Merchant Resource Center](https://esqa.moneris.com/mpg/index.php)
```
Username: demouser
Store ID: store5
Password: password
```
4. Find your transactions.
5. You should see (_example for VISA_)

- Card Verification 
<img width="1006" alt="Screenshot 2020-04-27 at 13 48 30" src="https://user-images.githubusercontent.com/8780680/80369290-5326b280-888e-11ea-9bfd-38467a959148.png">

- Purchase 1
<img width="1015" alt="Screenshot 2020-04-27 at 13 49 28" src="https://user-images.githubusercontent.com/8780680/80369353-6cc7fa00-888e-11ea-8e8d-02eeec3daa45.png">

- Purchase 2
<img width="1016" alt="Screenshot 2020-04-27 at 13 50 08" src="https://user-images.githubusercontent.com/8780680/80369377-75b8cb80-888e-11ea-8e75-cc143b4c3cde.png">
